tools/manifest/sourcefile.py

```diff
@@ -237,8 +238,12 @@ def url(self):
 
     @cached_property
     def hash(self):
-        with self.open() as f:
-            return hashlib.sha1(f.read()).hexdigest()
+        if not self._hash:
+            with self.open() as f:
+                content = f.read()
+            data = "blob " + str(len(content)) + "\0" + content
```

Comment:
So, I wonder if you profiled `"".join(["blob ", str(len(content)), "\0", content])` and `"blob %d\0%s" % (len(content), content)` vs this approach where the manifest isn't being generated from git (e.g. if you pass `--work` to the update). I don't know which is fastest, but it seems worth trying.

Reply:
Profiled them and seems like joins over tuples are the fastest of dealing with _very_ long strings. On my system, at least.

Reply:
And they'll definitely be quicker on PyPy, given the lack of refcounting means they can't do the ugly CPython hack.